src/formula/transform.rs

```diff
@@ -854,16 +930,111 @@ impl Theory {
     pub fn gnf(&self) -> Theory {
         let mut generator = SkolemGenerator::new();
         let formulas: Vec<Formula> = self.formulas.iter().flat_map(|f| f.gnf_with(&mut generator)).collect();
-        Theory::new(formulas)
+        Theory::new(compress_geometric(formulas))
     }
 }
 
+fn compress_geometric(formulas: Vec<Formula>) -> Vec<Formula> {
+    use itertools::Itertools;
+
+    formulas.into_iter().sorted_by(|f1, f2| { // sort sequents by their body
+        match (f1, f2) {
+            (Formula::Implies { left: f1, .. }, Formula::Implies { left: f2, .. }) => {
+                f1.cmp(f2)
+            }
+            _ => Equal
+        }
+    }).into_iter().coalesce(|f1, f2| { // merge the ones with the same body:
+        match f1 {
+            Formula::Implies { left: ref l1, right: ref r1 } => {
+                let l_vars = l1.free_vars();
+                let r_vars = r1.free_vars();
+                // compress sequents with no free variables that show up only in head:
+                if r_vars.iter().all(|rv| l_vars.iter().any(|lv| lv == rv)) {
+                    match f2 {
+                        Formula::Implies { left: ref l2, right: ref r2 } => {
+                            let l_vars = l2.free_vars();
+                            let r_vars = r2.free_vars();
+                            if r_vars.iter().all(|rv| l_vars.iter().any(|lv| lv == rv)) {
+                                if l1 == l2 {
+                                    let r = Formula::And { left: r1.clone(), right: r2.clone() };
+                                    Ok(Implies { left: l2.clone(), right: Box::new(r) })
```

Comment:
Early return for the Ok here so that you don't need the following else branches?

Reply:
The order of formulas in `Err` is important for the correctness of the algorithm. The idea is to walk through the items of a vector and merge the ones that can be merged. Sometimes I want to swap a formula with the one before it so the previous one can be merged with the next one.